models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+
+    V_.PSCInitialValues_[ i ] = 1.0 * numerics::e / P_.tau_syn_[ i ];
+    B_.spikes_[ i ].resize();
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::update( Time const& origin, const long from, const long to )
+{
+  // glif_lif_r_asc_a
+  const double dt = Time::get_resolution().get_ms();
+
+  // get model type
+  std::string model_str = P_.glif_model_;
+  std::transform( model_str.begin(), model_str.end(), model_str.begin(), ::tolower );
+  if ( nest::glif_psc::model_type_lu.find( model_str ) == nest::glif_psc::model_type_lu.end() )
+  {
+    throw BadProperty( "Bad glif model type string." );
+  }
+  long model_type = nest::glif_psc::model_type_lu[ model_str ];
+
+  double v_old = S_.U_;
+  double spike_component = 0.0;
+  double voltage_component = 0.0;
+  double th_old = S_.threshold_;
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+
+    // update threshold via exact solution of dynamics of spike component of
+    // threshold
+    // exact solution of dynamics of spike component of threshold for glif2/4/5
+    // models with "R"
+    if ( model_type == 2 || model_type == 4 || model_type == 5 )
+    {
+      spike_component = V_.last_spike_ * std::exp( -P_.b_spike_ * dt );
+    }
+    V_.last_spike_ = spike_component;
+
+    if ( V_.t_ref_remaining_ > 0.0 )
+    {
+      // While neuron is in refractory period count-down in time steps (since dt
+      // may change while in refractory) while holding the voltage at last peak.
+      V_.t_ref_remaining_ -= dt;
+      if ( V_.t_ref_remaining_ <= 0.0 )
+      {
+        // Neuron has left refractory period, reset voltage and after-spike
+        // current
+
+        // Reset ASC_currentsfor glif3/4/5 models with "ASC"
+        if ( model_type == 3 || model_type == 4 || model_type == 5 )
+        {
+          for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+          {
+            S_.ASCurrents_[ a ] =
+              P_.asc_amps_[ a ] + S_.ASCurrents_[ a ] * P_.r_[ a ] * std::exp( -P_.k_[ a ] * V_.t_ref_total_ );
+          }
+        }
+
+        // Reset voltage
+        if ( model_type == 1 || model_type == 3 )
+        {
+          // Reset voltage for glif1/3 models without "R"
+          S_.U_ = P_.V_reset_;
+        }
+        else
+        {
+          // Reset voltage for glif2/4/5 models with "R"
+          S_.U_ = P_.voltage_reset_a_ * S_.U_ + P_.voltage_reset_b_;
+
+          // reset spike component of threshold
+          V_.last_spike_ = V_.last_spike_ + P_.a_spike_;
+
+          // rest the global threshold (voltage component of threshold: stay the
+          // same)
+          S_.threshold_ = V_.last_spike_ + V_.last_voltage_ + P_.th_inf_;
+        }
+        // Check if bad reset
+        // TODO: Better way to handle?
+        if ( S_.U_ > S_.threshold_ )
+        {
+          printf(
```

Comment:
Throw an exception if this happens, e.g., `NumericalInstability`.

And can this actually happen? Could it make sense to allow for a minimal numerical error and in that case just set U to threshold?

Reply:
This problem not longer exists after reorganized the reset part of code.